keps/sig-cli/3638-built-in-command-shadowing/README.md

```diff
@@ -0,0 +1,685 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP allows external plugins to be used as subcommands of built-in commands when the subcommand does not exist.
+
+## Motivation
+
+There is a high demand for enriching some built-in kubectl commands with new subcommands. More specifically take an instance of create command, 
+even though it supports the creation of 18 resources(as of v1.25) with one line command execution(e.g. kubectl create job), 
+new requests to support other resources as well are occasionally filed. 
+These requests are understandable given the proliferation of CustomResourceDefinitions, but is not sustainable from the SIG CLI maintainers point of view. 
+Because that might bring about not only increase the size of the binary but also create a maintenance burden.
+
+This KEP proposes a viable alternative to users via plugins instead of creating static subcommands inside kubectl built-in commands.
+
+### Goals
+
+* Enable external plugin usage for `kubectl` built-in commands as subcommands, if subcommands does not exist.
+
+### Non-Goals
+
+* Implement a plugin
+
+## Proposal
+
+When user executes a subcommand of a built-in command, instead returning an "unknown command" error, `kubectl` will search this subcommand
+in external plugins with the name structure similar to current plugin search mechanism(e.g. `kubectl-create-foo`).
+
+```sh
+$ kubectl create foo
+(running kubectl-create-foo plugin)
+```
+
+If `kubectl` finds a match for the subcommand, it will execute this external plugin how it already does 
+for external plugins currently.
+
+Because of starting as alpha and this is an explicitly opted-in feature, shadowing functionality will be hidden behind the environment variable
+and will only be used after exporting;
+```sh
+export KUBECTL_ENABLE_ALPHA_CMD_SHADOW=true
+```
+
+### Notes/Constraints/Caveats
+
+This KEP in alpha and beta stage will only support shadowing for `kubectl create` to get feedback for other
+built-in commands as well. Currently, we know that there are requests for `kubectl create` command but we need to find out
+which other built-in commands will also be requested for this shadowing.
+
+### Risks and Mitigations
+
+Security considerations are the same with executing other external plugins and there is no additional risk.
```

Comment:
What happens if Kubernetes later adds a new `kubectl create foo` that happens to match a shadowed name? The new built-in takes precedence?

Do we have advice for users to safely migrate? Is it possible to detect built-in vs shadowed plugin?

This seems like a risk.

Reply:
Thanks for the review @BenTheElder. That is a really good point.  As you said the new built-in will take precedence. 

In order to prevent such confusions, I think we can add a mechanism to warn users if there is a plugin name in execution path. 

For example, if new subcommand is added `kubectl create foo`;

```sh
$ kubectl create foo
external plugin kubectl-create-foo is found but built-in subcommand will be executed. If you want to use plugin, please execute it directly.
```

The downside of this approach is, for all commands we always need to check external plugin exists or not.

Reply:
If users rely on plugin in their CI scripts and they really need to continue this plugin, even after the new built-in subcommand is added, we can propose a escape mechanism by introducing a `--force-plugin` flag.

```sh
$ kubectl create foo --force-plugin
I know there is a built-in subcommand, but some feature of the plugin is essential for my CI and still continue using external plugin
```

What do you think @soltysh ?

Reply:
Thanks Ben for pointing it out, that's very insightful observation and one of the reasons we're trying to limit the commands which will be allowed to be shadowed. I'm definitely supportive of having the force mechanism in place, but I'm inclined to say it should be hidden behind an env var rather than a flag, since that seems much more appropriate in those cases. @ardaguclu can you ensure this section is added and all the risks updated?

Reply:
I'm wondering if we should re-use the same env var for enablement and those exceptions, but I can't think of a reason why not, for now. We can re-evaluate that before beta.